src/ucm/mmap/install.c

```diff
@@ -62,72 +77,124 @@ static ucm_mmap_func_t ucm_mmap_funcs[] = {
+        /* generate UNMAP event */
+        UCM_FIRE_EVENT(events, UCM_EVENT_MMAP|UCM_EVENT_VM_MAPPED,
+                       cb, arg, p = mmap(p, ucm_get_page_size(), PROT_READ | PROT_WRITE,
+                                         MAP_FIXED | MAP_PRIVATE | MAP_ANONYMOUS, -1, 0));
+        UCM_FIRE_EVENT(events, UCM_EVENT_MUNMAP|UCM_EVENT_VM_UNMAPPED,
+                       cb, arg, munmap(p, ucm_get_page_size()));
     }
 
     if (events & (UCM_EVENT_SHMAT|UCM_EVENT_SHMDT|UCM_EVENT_VM_MAPPED|UCM_EVENT_VM_UNMAPPED)) {
-        shmid = shmget(IPC_PRIVATE, ucm_get_page_size(),
-                       IPC_CREAT | SHM_R | SHM_W);
-        p = shmat(shmid, NULL, 0);
-        shmdt(p);
+        shmid = shmget(IPC_PRIVATE, ucm_get_page_size(), IPC_CREAT | SHM_R | SHM_W);
+        if (shmid == -1) {
+            ucm_debug("shmget failed: %m");
+            return;
+        }
+
+        UCM_FIRE_EVENT(events, UCM_EVENT_SHMAT|UCM_EVENT_VM_MAPPED,
+                       cb, arg, p = shmat(shmid, NULL, 0));
+        UCM_FIRE_EVENT(events, UCM_EVENT_SHMAT|UCM_EVENT_VM_MAPPED|UCM_EVENT_VM_UNMAPPED,
+                       cb, arg, p = shmat(shmid, p, SHM_REMAP));
         shmctl(shmid, IPC_RMID, NULL);
+        UCM_FIRE_EVENT(events, UCM_EVENT_SHMDT|UCM_EVENT_VM_UNMAPPED,
+                       cb, arg, shmdt(p));
     }
 
     if (events & (UCM_EVENT_SBRK|UCM_EVENT_VM_MAPPED|UCM_EVENT_VM_UNMAPPED)) {
-        (void)sbrk(ucm_get_page_size());
-        (void)sbrk(-ucm_get_page_size());
+        UCM_FIRE_EVENT(events, UCM_EVENT_SBRK|UCM_EVENT_VM_MAPPED,
+                       cb, arg, (void)sbrk(ucm_get_page_size()));
+        UCM_FIRE_EVENT(events, UCM_EVENT_SBRK|UCM_EVENT_VM_UNMAPPED,
+                       cb, arg, (void)sbrk(-ucm_get_page_size()));
     }
 
     if (events & UCM_EVENT_MADVISE) {
-        p = mmap(NULL, ucm_get_page_size(), PROT_READ|PROT_WRITE,
-                 MAP_PRIVATE|MAP_ANON, -1, 0);
+        UCM_FIRE_EVENT(events, UCM_EVENT_MMAP|UCM_EVENT_VM_MAPPED, cb, arg,
+                       p = mmap(NULL, ucm_get_page_size(), PROT_READ|PROT_WRITE,
+                                MAP_PRIVATE|MAP_ANON, -1, 0));
         if (p != MAP_FAILED) {
-            madvise(p, ucm_get_page_size(), MADV_NORMAL);
-            munmap(p, ucm_get_page_size());
+            UCM_FIRE_EVENT(events, UCM_EVENT_MADVISE, cb, arg,
+                           madvise(p, ucm_get_page_size(), MADV_NORMAL));
+            UCM_FIRE_EVENT(events, UCM_EVENT_MUNMAP|UCM_EVENT_VM_UNMAPPED, cb, arg,
+                           munmap(p, ucm_get_page_size()));
         } else {
             ucm_debug("mmap failed: %m");
         }
     }
 }
 
+
+static void ucm_fire_mmap_empty_cb(int event, int pre_call, void *arg)
+{
+}
+
+void ucm_fire_mmap_events(int events)
+{
+    ucm_fire_mmap_events_with_cb(events, ucm_fire_mmap_empty_cb, NULL);
+}
+
+static void ucm_mmap_test_events_cb(int events, int pre_call, void *arg)
```

Comment:
per phone discussion, let's remove the callback and update the flags directly from the macro

Reply:
done